### PR TITLE
cmd/sync: add `check-all` and `check-new` options to compare checksums

### PR DESCRIPTION
**Pull request by SandyXSD**

### CODE REVIEW

File: pkg/sync/sync.go [hunk 1]
```diff
@@ -475,19 +557,25 @@ func producer(tasks chan<- object.Object, src, dst object.ObjectStorage, config
...
+		if dstobj == nil || obj.Key() < dstobj.Key() {
 			tasks <- obj
-		} else { // obj.Key() == dstobj.Key()
-			if config.DeleteSrc && obj.Size() == dstobj.Size() {
+		} else { // obj.key == dstobj.key
+			if config.ForceUpdate ||
+				(config.Update && obj.Mtime().After(dstobj.Mtime())) ||
+				(!config.Update && obj.Size() != dstobj.Size()) {
+				tasks <- obj
+			} else if config.Update { // obj.mtime <= dstobj.mtime
```
Comment (davies) [inline pkg/sync/sync.go:567]: We should verify the checksum when obj.mtime == dstobj.mtime

Comment (davies) [reply pkg/sync/sync.go:567]: source and destination may have different time resolution, should we round it to seconds?

File: pkg/sync/sync.go [hunk 2]
```diff
@@ -176,96 +179,151 @@ var bufPool = sync.Pool{
...
+	logger.Debugf("Copied permissions (%s:%s:%s) for %s in %s", fi.Owner(), fi.Group(), fi.Mode(), key, time.Since(start))
+}
 
+func doCheckSum(src, dst object.ObjectStorage, key string, equal *bool) error {
 	buf := bufPool.Get().(*[]byte)
 	defer bufPool.Put(buf)
-	if _, e = io.CopyBuffer(f, in, *buf); e != nil {
-		return e
+
+	srcIn, err := src.Get(key, 0, -1)
```
Comment (davies) [inline pkg/sync/sync.go:225]: limit the bandwidth and concurrent connections

File: pkg/sync/sync.go [hunk 3]
```diff
@@ -176,96 +179,151 @@ var bufPool = sync.Pool{
...
+	if err != nil {
+		return fmt.Errorf("src get: %s", err)
+	}
+	defer srcIn.Close()
+	srcH := md5.New()
+	if _, err = io.CopyBuffer(srcH, srcIn, *buf); err != nil {
```
Comment (davies) [inline pkg/sync/sync.go:231]: we read from two streams in the same time, should we just compare the bytes?

Or, we can compare the checksum block by block to find the different earlier.

Comment (davies) [reply pkg/sync/sync.go:231]: For larger objects, we can also compare them in parallel.

File: pkg/sync/sync.go [hunk 4]
```diff
@@ -309,74 +366,99 @@ func copyInParallel(src, dst object.ObjectStorage, obj object.Object) error {
...
+					deleteObj(src, key, false)
+				} else if config.Perms { // FIXME: && needCopyPerms(obj, dstobj)
+					copyPerms(dst, obj)
+					copied.Increment()
+				} else {
+					skipped.Increment()
+				}
+				break
+			}
+			// checkSum failed or not equal, copy the object
```
Comment (davies) [inline pkg/sync/sync.go:432]: For failed check, should we mark it as failed?

File: pkg/sync/sync.go [hunk 5]
```diff
@@ -309,74 +366,99 @@ func copyInParallel(src, dst object.ObjectStorage, obj object.Object) error {
...
+			if config.Dry {
+				logger.Infof("Will compare checksum for %s", key)
+				break
+			}
+			obj = obj.(*withSize).Object
+			if equal, err := checkSum(src, dst, key, obj.Size()); err == nil && equal {
+				if config.DeleteSrc {
+					deleteObj(src, key, false)
+				} else if config.Perms { // FIXME: && needCopyPerms(obj, dstobj)
+					copyPerms(dst, obj)
```
Comment (davies) [inline pkg/sync/sync.go:425]: check the permission (`dst.Head(key)`)before copy

File: pkg/sync/sync.go (hunk 4 above)
Comment (SandyXSD) [reply pkg/sync/sync.go:432]: Copying the object might fix the problem if error occurs in the destination side.

Comment (SandyXSD) [reply pkg/sync/sync.go:432]: Maybe it is safer to just mark it failed and let users to double check it.

File: pkg/sync/sync.go (hunk 3 above)
Comment (SandyXSD) [reply pkg/sync/sync.go:231]: Parallelism is not implemented because Get() returns a reader, not readerat

File: pkg/sync/sync.go [hunk 6]
```diff
@@ -171,101 +173,173 @@ func ListAll(store object.ObjectStorage, start, end string) (<-chan object.Objec
 
 var bufPool = sync.Pool{
 	New: func() interface{} {
-		buf := make([]byte, 32<<10)
+		buf := make([]byte, defaultPartSize)
```
Comment (SandyXSD) [inline pkg/sync/sync.go:176]: @davies not sure if this change is reasonable

Comment (davies) [reply pkg/sync/sync.go:176]: Don't, defaultPartSize is too large as an buffer

File: pkg/sync/sync.go (hunk 3 above)
Comment (davies) [reply pkg/sync/sync.go]: Get(key, offset, limit)

**Review by davies — APPROVED** [review]

**Pull request merged**